### PR TITLE
Use Prometheus to get queue length in auto-scaling

### PR DESCRIPTION
**Pull request by gchhablani**
This PR adds the Prometheus API call to query prometheus server and fetch the required attributes for calculating queue length in the auto-scaling script.

### CODE REVIEW

File: scripts/monitoring/auto_scale_workers.py [hunk 1]
```diff
@@ -2,22 +2,27 @@
...
 from auto_stop_workers import start_worker, stop_worker
+from prometheus_api_client import PrometheusConnect
+
+warnings.filterwarnings("ignore")
 
 utc = pytz.UTC
 
+NUM_PROCESSED_SUBMISSIONS = "num_processed_submissions"
+NUM_SUBMISSIONS_IN_QUEUE = "num_submissions_in_queue"
+PROMETHEUS_URL = "https://monitoring-staging.eval.ai/prometheus/"
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:16]: @gchhablani can we fetch the endpoint from a environment variable named `MONITORING_API_URL`.

File: scripts/monitoring/auto_scale_workers.py [hunk 2]
```diff
@@ -2,22 +2,27 @@
...
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:17]: Can we also read another environment variable which is named `ENV` with values `dev` or `prod`. If the environment is prod we should only perform auto scaling on the following challenge queues and ignore the rest:

1. VQA challenge: `vqa-challenge-2021-744-production-8228992d-b6c6-4e18-b6c4-ce2e8dbe9a57`
2. KILT challenge: `kilt-69d368bc-a5e8-4952-bd68-21c3ba61eb65`
3. NuScenes challenge: `nuscenes-detection-challenge-510c8c6d-a0d2-40bd-95dd-b7c8ea593d03`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/monitoring/auto_scale_workers.py [hunk 3]
```diff
@@ -2,22 +2,34 @@
...
 
-utc = pytz.UTC
+warnings.filterwarnings("ignore")
 
+utc = pytz.UTC
+NUM_PROCESSED_SUBMISSIONS = "num_processed_submissions"
+NUM_SUBMISSIONS_IN_QUEUE = "num_submissions_in_queue"
+PROMETHEUS_URL = os.getenv(
+    "MONITORING_API_URL"
+)  # "https://monitoring-staging.eval.ai/prometheus/"
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:17]: Can you remove the comment? Also, can you set the default value of env variable to be staging endpoint? You can use `os.environ.get` to do that. See this example https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/aws_utils.py#L40

File: scripts/monitoring/auto_scale_workers.py [hunk 4]
```diff
@@ -2,22 +2,34 @@
...
+PROD_CHALLENGE_QUEUES = [
+    "vqa-challenge-2021-744-production-8228992d-b6c6-4e18-b6c4-ce2e8dbe9a57",
+    "kilt-69d368bc-a5e8-4952-bd68-21c3ba61eb65",
+    "nuscenes-detection-challenge-510c8c6d-a0d2-40bd-95dd-b7c8ea593d03",
+]
+ENV = os.getenv("ENV")
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:23]: same here

File: scripts/monitoring/auto_scale_workers.py [hunk 5]
```diff
@@ -2,22 +2,34 @@
...
 
 # Eval AI related credentials
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:25]: remove the comment

File: scripts/monitoring/auto_scale_workers.py [hunk 6]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
-    return (
-        queue.attributes["ApproximateNumberOfMessages"]
-        + queue.attributes["ApproximateNumberOfMessagesNotVisible"]
-    )
+        num_submissions_in_queue = int(
+            prom.custom_query(
+                f"num_submissions_in_queue{{queue_name='{queue_name}'}}"
+            )[0]["value"][1]
+        )
+    except Exception as _e:  # noqa: F841
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:65]: We are not using `_e` anywhere so just change to `except Exception`

File: scripts/monitoring/auto_scale_workers.py [hunk 7]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
-    if queue_name == "":
-        raise ValueError("Queue Name cannot be empty")
+def get_queue_length(queue_name):
+    try:
+        num_processed_submissions = int(
+            prom.custom_query(
+                f"num_processed_submissions{{queue_name='{queue_name}'}}"
+            )[0]["value"][1]
+        )
+    except Exception as _e:  # noqa: F841
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:55]: We are not using `_e` anywhere so just change to except Exception

File: scripts/monitoring/auto_scale_workers.py [hunk 8]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
-        client = boto3.client(
-            resource,
-            region_name=aws_keys["AWS_REGION"],
-            aws_access_key_id=aws_keys["AWS_ACCESS_KEY_ID"],
-            aws_secret_access_key=aws_keys["AWS_SECRET_ACCESS_KEY"],
-        )
-        return client
-    except Exception as e:
-        raise e
+    print(queue_name)
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:74]: please remove this print statement

File: scripts/monitoring/auto_scale_workers.py [hunk 9]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
+            "Unable to get the queue length for challenge ID: {}. Skipping.".format(
+                challenge["id"]
+            )
+        )
+        return
+
+    if queue_length == 0:
+        if challenge["workers"] is not None and int(challenge["workers"]) > 0:
+            # Worker > 0 and Queue = 0 - Stop
+            # stop worker
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:92]: please remove these comments

File: scripts/monitoring/auto_scale_workers.py [hunk 10]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
+            stop_worker(challenge["id"])
+            print("Stopped worker for challenge: {}".format(challenge["id"]))
+        else:
+            # Worker = 0 and Queue = 0
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:96]: same here

File: scripts/monitoring/auto_scale_workers.py [hunk 11]
```diff
@@ -33,101 +45,100 @@ def get_challenges():
...
+            # start worker
+            start_worker(challenge["id"])
+            print("Started worker for challenge: {}".format(challenge["id"]))
+        else:
+            # Worker > 0 and Queue > 0
+            print(
+                "Existing workers and pending queue messages found for challenge: {}. Skipping.".format(
+                    challenge["id"]
+                )
+            )
```
Comment (Ram81) [inline scripts/monitoring/auto_scale_workers.py:115]: remove comments

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/monitoring/auto_scale_workers.py (hunk 4 above)
Comment (gchhablani) [reply scripts/monitoring/auto_scale_workers.py:23]: Setting the default to "dev".

File: scripts/monitoring/auto_scale_workers.py (hunk 6 above)
Comment (gchhablani) [reply scripts/monitoring/auto_scale_workers.py:65]: Resolved.

File: scripts/monitoring/auto_scale_workers.py (hunk 7 above)
Comment (gchhablani) [reply scripts/monitoring/auto_scale_workers.py:55]: Resolved.

File: scripts/monitoring/run_auto_scale_workers.sh [hunk 12]
```diff
@@ -21,4 +21,4 @@ fi
 export AUTH_TOKEN=${auth_token};
 export API_HOST_URL=${api_host_url};
 
-python ${path}/scripts/monitoring/auto_scale_workers.py
+/home/ubuntu/venv/bin/python ${path}/scripts/monitoring/auto_scale_workers.py
```
Comment (Ram81) [inline scripts/monitoring/run_auto_scale_workers.sh:24]: @gchhablani can you also set `MONITORING_API_URL` in the shell script? This way crontab won't be able to find the prometheus API URL

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Review by Ram81 — APPROVED** [review]
LGTM

**Review by Ram81 — APPROVED** [review]
LGTM

**Pull request merged**